### PR TITLE
Backend Base class: support context

### PR DESCRIPTION
**Pull request by aesteve-rh**
Implement `__enter__` and `__exit__` methods in the
backend base class so that all storage sub classes
support context management.

Add tmpdir method to create a temporary directory
in the current mounted filesystem that can be used
as a context manager.

Signed-off-by: Albert Esteve <aesteve@redhat.com>

### CODE REVIEW

File: userstorage/backend.py [hunk 1]
```diff
@@ -76,7 +76,7 @@ def setup(self):
         the tests. If you use unittest based tests, should be called in
         setUp().
 
-        Subclass should implemnet if needed.
+        Subclass should implement if needed.
```
Comment (nirs) [inline userstorage/backend.py]: actually override is a better term.

**Review by nirs — COMMENTED** [review]
Looks good, but needs tests using the context manager interface.

**Comment by aesteve-rh** [comment]
I draft the PR for now. There is a test that is failing in current master. I will open a new PR enabling the pipelines and removing the failing test, and then we can integrate this one.

File: test/consume_test.py [hunk 2]
```diff
@@ -44,11 +43,9 @@ def user_loop(request):
...
 )
 def user_mount(request):
-    backend = validate_backend(request.param)
-    if backend.fstype == "ext2":
-        pytest.xfail("block size detection broken since kernel 5.5")
-    storage = validate_backend(request.param)
-    with storage as backend:
+    with request.param as backend:
+        if backend.fstype == "ext2":
+            pytest.xfail("block size detection broken since kernel 5.5")
```
Comment (nirs) [inline test/consume_test.py]: This fails all tests with this backend, while the only issue is not being able to detect block size.

Better skip only the test for detecting block size.

File: userstorage/backend.py [hunk 3]
```diff
@@ -95,6 +97,8 @@ def __enter__(self):
         Setup the current storage and return it.
         """
         self.setup()
+        if not self.exists():
+            pytest.xfail(f"backend {self} not available")
```
Comment (nirs) [inline userstorage/backend.py:100]: setup() will fail if the backend does not exist. The check should be before calling setup. The purpose of the test is to xfail tests using this storage, when the storage is defined with required=False.

You need a test verifying this - create a backend instance without calling create(), and then using it as context manager and verifying that the test is xfailed (not sure how this can be done) or least does not fail.

File: userstorage/mount.py [hunk 4]
```diff
@@ -80,6 +81,13 @@ def setup(self):
             else:
                 os.remove(path)
 
+    def tmpdir(self):
+        """
+        Create and return a TemporaryDirectory in the current
+        mounted filesystem. Can be used as a context manager.
+        """
+        return tempfile.TemporaryDirectory(dir=self.path)
```
Comment (nirs) [inline userstorage/mount.py:89]: This feature is not tested - can you add test veriying that we got a new directory inside the mount, and the directory is deleted after the test?

File: setup.py [hunk 5]
```diff
@@ -7,7 +7,7 @@
 
 setuptools.setup(
     name="userstorage",
-    version="0.4",
+    version="0.5",
```
Comment (nirs) [inline setup.py]: Lets split this to another pr.

**Review by nirs — CHANGES_REQUESTED** [review]

**Comment by aesteve-rh** [comment]
With the current status, if the example_config storage is not created, some tests would fail (not xfail), since some backends are actually required in the configuration.

I think it is nice to have it mixed to showcase that it needs to set required to false in order to xfail. But if we want to keep the old behavior, where we xfailed every time, we should probably update the configuration.

File: test/consume_test.py [hunk 6]
```diff
@@ -80,6 +78,8 @@ def test_loop_device(user_loop):
 
 def test_mount(user_mount):
     assert os.path.isdir(user_mount.path)
+    if user_mount.fstype == "ext2":
+        pytest.xfail("block size detection broken since kernel 5.5")
```
Comment (nirs) [inline test/consume_test.py]: Why skip the entire test when the only thing we cannot test is line 88?

We can do:

    if user_mount.fstype == "ext2":
        log.warning(...)
    else:
        assert detect_block_size(filename) == user_mount.sector_size

File: test/consume_test.py [hunk 7]
```diff
@@ -85,6 +86,15 @@ def test_mount(user_mount):
     assert detect_block_size(filename) == user_mount.sector_size
 
 
+def test_mount_tmpdir(user_mount):
+    with user_mount.tmpdir() as tmpdir:
+        assert os.path.isdir(tmpdir)
+        # Check that tmpdir is relative to the mount path
+        assert Path(user_mount.path) in Path(tmpdir).parents
+
+    assert not os.path.exists(tmpdir)
```
Comment (nirs) [inline test/consume_test.py:90]: Nice!

File: test/provision_test.py [hunk 8]
```diff
@@ -74,3 +74,18 @@ def test_delete_twice(cleanup):
 
     for b in BACKENDS.values():
         assert not b.exists()
+
+
+@pytest.mark.sudo
+@pytest.mark.xfail(reason="Test end with xfail", strict=True)
```
Comment (nirs) [inline test/provision_test.py]: With this this test will always xfail, while we want it to pass.

File: test/provision_test.py [hunk 9]
```diff
@@ -74,3 +74,18 @@ def test_delete_twice(cleanup):
...
+    storage = BACKENDS["block-512"]
+    # Setup the storage without create raises exception
+    with pytest.raises(subprocess.CalledProcessError):
+        with storage:
+            pass
+
+    # Setup a non-required storage without create is xfail
+    storage.required = False
+    with storage:
+        pass
```
Comment (nirs) [inline test/provision_test.py]: I think that pytest.xfail() raises some exception. Can we test that this block raises the expected exception?

File: userstorage/backend.py [hunk 10]
```diff
@@ -94,6 +96,8 @@ def __enter__(self):
         """
         Setup the current storage and return it.
         """
+        if not self.exists() and not self.required:
```
Comment (nirs) [inline userstorage/backend.py]: Why check if the storage is required? maybe add a comment explaining the check?

**Review by nirs — CHANGES_REQUESTED** [review]

File: test/provision_test.py (hunk 8 above)
Comment (aesteve-rh) [reply test/provision_test.py]: Initially, my idea with this mark was to make sure it will xfail, hence the `strict=True`. That way, if it did not execute `pytest.xfail`, it would actually fail.

But let's try to make it pass instead.

File: test/provision_test.py (hunk 9 above)
Comment (aesteve-rh) [reply test/provision_test.py]: Yes, it raises `XFailed` (https://github.com/pytest-dev/pytest/blob/main/src/_pytest/outcomes.py#L253), which is not exposed in the public pytest module.

I will check that the test raises XFailed, importing the exception from the private `_pytest` module.

File: userstorage/backend.py (hunk 10 above)
Comment (aesteve-rh) [reply userstorage/backend.py]: Inferred it from this comment: https://github.com/nirs/userstorage/pull/11#discussion_r969593132

But now that I recheck with the code, I think you meant it as in required storage will fail during `create`, and non-required storage will xfail during `setup`, so I will remove `required` from this condition.

**Review by nirs — APPROVED** [review]
